docs/user_docs/guides/abstraction.md

```diff
@@ -0,0 +1,529 @@
+
+We define our application once with the `App` schema and then use the KCL CLI to translate it to multiple platforms, such as Docker Compose or Kubernetes. The application model aims to reduce developer toil and cognitive load by only having to define a single KCL file that works across multiple platforms, and it is designed to be applied to multiple environments to reduce the amount of configuration.
+
+Let's learn how to do this.
+
+#### Define an Application Model
+
+Firstly, we define an application model using the schema in the KCL file (app.k).
+
+```python
+schema App:
+    """The application model."""
+    name: str
+    replicas: int = 1
+    labels?: {str:str}
+    service?: Service
+    containers?: {str: Container}
+
+schema Service:
+    """The service model."""
+    $type?: str
+    ports: [Port]
+
+schema Port:
+    """The port model."""
+    port: int
+    protocol: "TCP" | "UDP" | "SCTP" = "TCP"
+    targetPort?: int | str
+
+schema Container:
+    """The container model."""
+    image: str
+    command?: [str]
+    args?: [str]
+    env?: [Env]
+    volumes?: [Volume]
+    resources: Resource
+    ports: [ContainerPort]
+
+schema ContainerPort:
+    """The container port model."""
+    name?: str
+    protocol: "TCP" | "UDP" | "SCTP" = "TCP"
+    containerPort: int
+
+    check:
+        1 <= containerPort <= 65535, "containerPort must be between 1 and 65535, inclusive"
+
+schema Env:
+    name: str
+    value: str
+
+schema Volume:
+    source: str
+    path: str
+    target: str
+    readOnly?: bool = False
+
+schema Resource:
+    limits?: {str:}
+    requests?: {str:}
+```
+
+#### Convert the Application Model into Docker Compose Config
+
+Define the docker compose render file (docker_compose_render.k).
+
+```python
+import manifests
+
+import .app
+
+# Convert the `App` model into the docker compose config.
+dockerComposeRender = lambda a: app.App {
+    # NOTE: Only one container per workload can be defined for compose.
+    #       All other containers will be ignored by this tool.
+    c: app.Container = [c for _, c in a.containers][0] if a.containers else {}
+    # Construct the docker compose config.
+    [{
+        services = {
+            "${a.name}" = {
+                image = c.image
+                entrypoint = c.command
+                environment = c.env
+                volumes = c.volumes
+                ports = [{
+                    published = p.port
+                    target = p.targetPort or p.port
+                    protocol = p.protocol
+                } for p in a.service?.ports]
+            }
+        }
+    }]
+}
+
+# Process multiple app instances and output them as the docker compose config.
+# Use the `sum` function to flatten the resource list.
+manifests.yaml_stream(sum([dockerComposeRender(a) for a in app.App.instances()], []))
+```
+
+#### Convert the Application Model into Kubernetes Deployment and Service Manifests
```

Comment:
Split approach documents from the sample documents. Place manifests output first. Place render code lastly.